helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeSnapshot.java

```diff
@@ -113,4 +123,26 @@
   ClusterConfig getClusterConfig() {
     return _clusterConfig;
   }
+
+  // Trim the IdealState to exclude any controller modified information.
+  private IdealState trimIdealState(IdealState originalIdealState) {
+    // Clone the IdealState to avoid modifying the objects in the Cluster Data Cache, which might
+    // be used by the other stages in the pipeline.
+    IdealState trimmedIdealState = new IdealState(originalIdealState.getRecord());
+    switch (originalIdealState.getRebalanceMode()) {
+      case FULL_AUTO:
+        // For FULL_AUTO resources, both map fields and list fields are not considered as data input
+        // for the controller.
```

Comment:
Also add:
The controller will write to mapFields and listFields meaning they will be overwritten by the controller...